dnf5/commands/provides/provides.cpp

```diff
@@ -0,0 +1,127 @@
+}
+
+void ProvidesCommand::set_argument_parser() {
+    auto & ctx = get_context();
+    auto & parser = ctx.get_argument_parser();
+
+    auto & cmd = *get_argument_parser_command();
+    cmd.set_description("Find what package provides the given value");
+
+    auto * keys = parser.add_new_positional_arg("specs", ArgumentParser::PositionalArg::AT_LEAST_ONE, nullptr, nullptr);
+    keys->set_description("List of package specs to query");
+    keys->set_parse_hook_func(
+        [this]([[maybe_unused]] ArgumentParser::PositionalArg * arg, int argc, const char * const argv[]) {
+            for (int i = 0; i < argc; ++i) {
+                pkg_specs.emplace_back(argv[i]);
+            }
+            return true;
+        });
+    keys->set_complete_hook_func([&ctx](const char * arg) { return match_specs(ctx, arg, false, true, true, false); });
+    cmd.register_positional_arg(keys);
+}
+
+void ProvidesCommand::configure() {
+    auto & context = get_context();
+    context.set_load_system_repo(true);
+    context.base.get_config().get_optional_metadata_types_option().add_item(
+        libdnf5::Option::Priority::RUNTIME, libdnf5::METADATA_TYPE_FILELISTS);
+    context.set_load_available_repos(Context::LoadAvailableRepos::ENABLED);
+}
+
+void ProvidesCommand::run() {
+    auto & ctx = get_context();
+    bool any_match = false;
+    std::set<std::string> unmatched_specs;
+
+    for (auto & spec : pkg_specs) {
+        libdnf5::rpm::PackageSet result_pset(ctx.base);
+        libdnf5::rpm::PackageQuery full_package_query(ctx.base);
+        auto provides_query = full_package_query;
+        auto filename_query = full_package_query;
+        int match = 0;
+        // if the spec starts with "/" assume it's a provide by filename and skip this query
+        if (!(spec.rfind("/", 0) == 0)) {
+            provides_query.filter_provides(std::vector<std::string>({spec}), libdnf5::sack::QueryCmp::GLOB);
+            if (!provides_query.empty()) {
+                full_package_query = provides_query;
+                match = libdnf5::cli::output::ProvidesMatchedBy::PROVIDES;
+            }
+        }
+        // compatibility for packages that didn't do UsrMove
+        if ((spec.rfind("/bin/", 0) == 0) || (spec.rfind("/sbin/", 0) == 0)) {
+            spec.insert(0, "/usr");
+        }
+        filename_query.filter_file(std::vector<std::string>({spec}), libdnf5::sack::QueryCmp::GLOB);
+        if (!filename_query.empty()) {
+            full_package_query = filename_query;
+            match = libdnf5::cli::output::ProvidesMatchedBy::FILENAME;
+        }
+
+        result_pset |= full_package_query;
+
+        for (auto package : result_pset) {
+            if (match != 0) {
+                libdnf5::cli::output::print_provides_table(package, spec.c_str(), match);
+                any_match = true;
+            } else {
+                unmatched_specs.insert(spec);
+            }
+        }
+        if (!unmatched_specs.empty() && any_match) {
```

Comment:
So, what if we have a list with several specs, and only the last one is actually matched, but the previous ones having unmatched specs. In this case, the error output is not printed for the unmatched ones. Additionally, if you were to reorder that list of arguments with the matched one first, then the output would be present. However, the ordering should not affect this, right?

Reply:
```
$ ./build/dnf5/dnf5 -q provides asd asdd wget 
wget-1.21.3-7.fc39.aarch64 : A utility for retrieving files using the HTTP or FTP protocols
Repo         : @System
Matched From : 
Provide      : wget = 1.21.3-7.fc39

wget-1.21.3-7.fc39.aarch64 : A utility for retrieving files using the HTTP or FTP protocols
Repo         : fedora
Matched From : 
Provide      : wget = 1.21.3-7.fc39

No matches found for asd.
No matches found for asdd.
If searching for a file, try specifying the full path or using a wildcard prefix ("*/") at the beginning.
$ echo $?
1
```

It shouldn't. this is what is happening in your example. did I get it right?

Reply:
this should be fixed, I'll update the tests to track the wanted behavior